File: PartitionList/PartitionList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        if head is None:
            return head
        node = head
        less_head = ListNode(0)
        greater_head = ListNode(0)
        less = less_head
        greater = greater_head
        while node is not None:
            list_next = node.next
            if node.val < x:
                less.next = node
                less = less.next
                less.next = None
            else:
                greater.next = node
                greater = greater.next
                greater.next = None
            node = list_next
        less.next = greater_head.next
        return less_head.next

File: PartitionList/PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if(head == nullptr || head->next == nullptr)
        	return head;
        ListNode* new_head = new ListNode(0);
        ListNode* pre = new_head;
        new_head->next = head;
        ListNode* cur = head;
        ListNode* tail = head;
        int count = 0;
        while(tail->next) {
        	tail = tail->next;
        	count++;
        }
        while(count >= 0) {
        	if(cur->val >= x && cur != tail) {
        		pre->next = cur->next;
        		tail->next = cur;
        		cur->next = nullptr;
        		tail = cur;
        		cur = pre->next;
        	}
        	else {
        		pre = cur;
        		cur = cur->next;
        	}
        	count--;
        }
        return new_head->next;
    }
};